mujoco_warp/_src/smooth.py

```diff
@@ -2038,10 +2048,141 @@ def _transmission(
 
         actuator_moment_out[worldid, actid, i] = moment
   else:
-    # TODO(team): slidercrank, body
     wp.printf("unhandled transmission type %d\n", trntype)
 
 
+@wp.kernel
+def _transmission_body_moment(
+  # Model:
+  opt_cone: int,
+  body_parentid: wp.array(dtype=int),
+  body_rootid: wp.array(dtype=int),
+  dof_bodyid: wp.array(dtype=int),
+  geom_bodyid: wp.array(dtype=int),
+  actuator_trnid: wp.array(dtype=wp.vec2i),
+  actuator_trntype_body_adr: wp.array(dtype=int),
+  # Data in:
+  ncon_in: wp.array(dtype=int),
+  subtree_com_in: wp.array2d(dtype=wp.vec3),
+  cdof_in: wp.array2d(dtype=wp.spatial_vector),
+  contact_dist_in: wp.array(dtype=float),
+  contact_pos_in: wp.array(dtype=wp.vec3),
+  contact_frame_in: wp.array(dtype=wp.mat33),
+  contact_includemargin_in: wp.array(dtype=float),
+  contact_dim_in: wp.array(dtype=int),
+  contact_geom_in: wp.array(dtype=wp.vec2i),
+  contact_efc_address_in: wp.array2d(dtype=int),
+  contact_worldid_in: wp.array(dtype=int),
+  efc_J_in: wp.array2d(dtype=float),
+  # Data out:
+  actuator_moment_out: wp.array3d(dtype=float),
+  actuator_trntype_body_ncon_out: wp.array2d(dtype=int),
+):
+  trnbodyid, conid, dofid = wp.tid()
+  actid = actuator_trntype_body_adr[trnbodyid]
+  bodyid = actuator_trnid[actid][0]
+
+  if conid >= ncon_in[0]:
+    return
+
+  worldid = contact_worldid_in[conid]
+
+  # get geom ids
+  geom = contact_geom_in[conid]
+  g1 = geom[0]
+  g2 = geom[1]
+
+  # contact involving flex, continue
+  if g1 < 0 or g2 < 0:
+    return
+
+  # get body ids
+  b1 = geom_bodyid[g1]
+  b2 = geom_bodyid[g2]
+
+  # irrelevant contact, continue
+  if b1 != bodyid and b2 != bodyid:
+    return
+
+  contact_exclude = int(contact_dist_in[conid] >= contact_includemargin_in[conid])
+
+  # mark contact normals in efc_force
+  if contact_exclude == 0:
+    if dofid == 0:
```

Comment:
do we need this `if dofid == 0:` in both branches?

seems like we could move it up out of the if statements?